### PR TITLE
GUACAMOLE-192: Double and triple click function

### PR DESCRIPTION
**Pull request by Saiyeux**
Two files added to /src/protocols/ssh and small changes to input.c. Now we can double click a character and select the whole word. Also triple to the whole line.

### CODE REVIEW

File: src/protocols/ssh/input.c [hunk 1]
```diff
@@ -58,7 +174,7 @@ int guac_ssh_user_key_handler(guac_user* user, int keysym, int pressed) {
 
     /* Report key state within recording */
     if (ssh_client->recording != NULL)
-        guac_recording_report_key(ssh_client->recording,
+        guac_common_recording_report_key(ssh_client->recording,
```
Comment (mike-jumper) [inline src/protocols/ssh/input.c:177]: This should remain `guac_recording_report_key()`. There is no `guac_common_recording_report_key()` function.

File: src/protocols/ssh/input.c [hunk 2]
```diff
@@ -88,8 +204,7 @@ int guac_ssh_user_size_handler(guac_user* user, int width, int height) {
     if (ssh_client->term_channel != NULL) {
         pthread_mutex_lock(&(ssh_client->term_channel_lock));
         libssh2_channel_request_pty_size(ssh_client->term_channel,
-                guac_terminal_get_columns(terminal),
-                guac_terminal_get_rows(terminal));
+                terminal->term_width, terminal->term_height);
```
Comment (mike-jumper) [inline src/protocols/ssh/input.c:207]: This shouldn't change to `terminal->term_width`, etc. which are now internal members of the terminal struct. The getter functions are the public interface.

File: src/protocols/ssh/input.c [hunk 3]
```diff
@@ -31,6 +33,20 @@
...
+
+int state = 0;
+long long start_1, start_2, start_3 = 0;
+long long end_1, end_2, end_3 = 0;
+
+long long getSystemTime() {
+    struct timeb t;
+    ftime(&t);
+    return 1000 * t.time + t.millitm;
+}
```
Comment (mike-jumper) [inline src/protocols/ssh/input.c:48]: You do not need to declare your own function for this - libguac provides `guac_timestamp_current()`.

File: src/protocols/ssh/input.c [hunk 4]
```diff
@@ -31,6 +33,20 @@
 
 #include <pthread.h>
 
+#include <stdio.h>
+#include <sys/timeb.h>
+
+
+int state = 0;
+long long start_1, start_2, start_3 = 0;
+long long end_1, end_2, end_3 = 0;
```
Comment (mike-jumper) [inline src/protocols/ssh/input.c:42]: Global state should be avoided where possible. This would make more sense within the state representation of the internal terminal struct, and needs to be documented.

File: src/protocols/ssh/click.h [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    guac_socket* socket; 
+    
+    guac_layer* select_layer;
+
+} guac_click;
+
+/* Display selection effect */
+void guac_click_draw_select(guac_click* click);
+
+void guac_click_draw_blank(guac_click* click);
```
Comment (mike-jumper) [inline src/protocols/ssh/click.h:51]: 1. All functions, parameters, return values, structures, and members need to be documented.
2. This would make more sense as part of the terminal, rather than being SSH-specific. In that case, each of these functions and structures should be renamed to be within the `guac_terminal_` namespace.

File: src/protocols/ssh/click.c [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+                /* Draw selection */
+                guac_click_draw_select(moved_selection);
+
+                return;
+
+            }
+
+            /* Marks */
+
+            else if ((sentinel > 32 && sentinel < 48) || (sentinel > 57 && sentinel < 65) || (sentinel > 90 && sentinel < 95) || (sentinel > 122 && sentinel < 127) ||(sentinel == 96)){
```
Comment (mike-jumper) [inline src/protocols/ssh/click.c:427]: There are a lot of magic numbers here and elsewhere in this function. Is there a better way to write this that would make things more readable and verifiable?

File: src/protocols/ssh/click.c [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+            /* Copy to clipboard */
+            guac_click_select_blank(selection);
+
+            /* Draw selection */
+            guac_click_draw_select(selection);
+
+            return;
+        }
+
+        else if (head > 0 && tail == width - 1){
```
Comment (mike-jumper) [inline src/protocols/ssh/click.c:392]: Please watch out for the spacing - there should be a space between the parenthesis and the opening brace. Same elsewhere between `if`/`while` and the opening paren.

File: src/protocols/ssh/click.c [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+void guac_double_click(guac_click* click){
+
+    /* Set selection variables */
+    guac_terminal* term = click->term;
+    int row = click->select_row;
+    int col = click->select_col;
+    printf("Row: %d, Col: %d\n", row, col);
+    int sentinel = term->display->operations[row * term->display->width + col].character.value;
+
+    /* Determination */
```
Comment (mike-jumper) [inline src/protocols/ssh/click.c:330]: What do you mean by this?

File: src/protocols/ssh/click.c [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+
+    /* Words */
```
Comment (mike-jumper) [inline src/protocols/ssh/click.c:332]: It's unclear to me what is meant by this comment. Can you rephrase to clarify?

File: src/protocols/ssh/click.c [hunk 10]
```diff
@@ -0,0 +1,457 @@
...
+        guac_click_select_word(selection);
+
+        /* Draw selection */
+        guac_click_draw_select(selection);
+
+        return;
+
+    }
+
+    /* Marks */
```
Comment (mike-jumper) [inline src/protocols/ssh/click.c:349]: Same here - what do you mean by "marks"?

File: src/protocols/ssh/click.c [hunk 11]
```diff
@@ -0,0 +1,457 @@
...
+    int flag = sentinel;
+
+    /* Get head */
+    while ((flag == 0 || flag == 32) && (head >= 0 && head <= term->display->width)){
+        flag = term->display->operations[row * term->display->width + head].character.value;
+        head--;
+    }
+
+    /* Reset flag */
+    flag = sentinel;
```
Comment (mike-jumper) [inline src/protocols/ssh/click.c:300]: Please rephrase to capture the intent of what's happening here, rather than the literal English equivalent of the relevant statement.

File: src/protocols/ssh/click.c [hunk 12]
```diff
@@ -0,0 +1,457 @@
...
+    
+    int row = click->select_row;
+    int col = click->select_col;
+    int head = click->select_head;
+    int tail = click->select_tail;
+
+    int sentinel = term->display->operations[row * term->display->width + col].character.value;
+    int flag = sentinel;
+
+    /* Get head */
```
Comment (mike-jumper) [inline src/protocols/ssh/click.c:293]: What do you mean by "head"?

File: src/protocols/ssh/click.c [hunk 13]
```diff
@@ -0,0 +1,457 @@
...
+            
+    head += 2;
+    tail -= 2;
+
+    int h = term->display->operations[row * term->display->width + head - 1].character.value;
+    if (head == 1 && ((h > 64 && h < 91) || (h > 96 && h < 123) || (h > 47 && h < 58)  || (h == 95)))
+        head = 0;
+
+    printf("Head: %d\n", head);
+    printf("Tail: %d\n", tail);
```
Comment (mike-jumper) [inline src/protocols/ssh/click.c:271]: Please do not include debug `printf()` calls.

File: src/protocols/ssh/click.c [hunk 14]
```diff
@@ -0,0 +1,457 @@
...
+    0x00, 0x80, 0xFF, 0x60);
+
+}
+
+void guac_click_draw_blank(guac_click* click){
+
+    guac_protocol_send_rect(click->socket, click->select_layer, 0, 0, 1, 1);
+
+    guac_protocol_send_cfill(click->socket, GUAC_COMP_SRC, click->select_layer,
+    0x00, 0x00, 0x00, 0x00);
```
Comment (mike-jumper) [inline src/protocols/ssh/click.c:56]: Why does the double/triple click behavior need its own dedicated selection redraw code, its own dedicated select layer, etc.? Can the existing terminal selection code not be used as-is?

**Review by mike-jumper — CHANGES_REQUESTED** [review]
Thanks - please take a look over this initial review and apply the feedback broadly. I have not yet read through the full internals of these proposed changes, however it looks like there is some major restructuring needed first. There are a few points with respect to style, documentation, etc. which I've made in individual locations but which apply elsewhere here. For the sake of brevity, I've only made those comments once.

**Comment by Saiyeux** [comment]
Thanks for these replies, I will read each suggestions and make reasonable changes.

**Comment by Saiyeux** [comment]
The newest commit with SHA of fa3d54b932d1a58082d7e06dbc6e810b63254775 should meet these needs mentioned above. I tried to resolve these problems by adding members to terminal structure. If this request is still available for GUACAMOLE-192, please let me know other potential hazards or format problems. Thanks for replying. @mike-jumper

**Comment by necouchman** [comment]
@Saiyeux : Actual code changes aside, you have a couple of issues with your pull request:
* You have a merge commit that should not be present. You'll need to rebase off the current master and get rid of that.
* You have have merge conflicts that need to be resolved with the current master before these changes will commit.
* All of your commit messages are exactly the same, and do not provide good descriptions of the nature of the commit. You'll need to update the commit messages, and possibly squash commits together where it makes sense.

**Comment by necouchman** [comment]
Closing as these changes are more or less covered by the recently-merged #504 .
